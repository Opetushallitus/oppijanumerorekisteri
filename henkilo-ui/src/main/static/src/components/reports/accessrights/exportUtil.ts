import { unparse } from 'papaparse';
import moment from 'moment';
import PropertySingleton from '../../../globals/PropertySingleton';
import type { TableHeading } from '../../../types/react-table.types';

const BOM = '\ufeff';
const DELIMITER = ';';

const isDate = (value) => String(value).match(/^\d{4}-\d{2}-\d{2}/);

const formatDate = (date: Date) => moment(date).format(PropertySingleton.state.PVM_MOMENT_FORMAATTI);

const format: (any) => any = (value) => (isDate(value) ? formatDate(value) : value);

export const convertToCSV = (
    data: Record<string, any>[],
    columns: TableHeading[],
    translate: (string) => string
): Record<string, string>[] => {
    const fields: Record<string, string> = columns.reduce(
        (acc, curr) => ({
            ...acc,
            [curr.key]: curr.label,
        }),
        {}
    );
    return data.map((row) =>
        Object.entries(fields).reduce(
            (acc, [key, label]) => ({
                ...acc,
                [translate(label)]: format(row[key]),
            }),
            {}
        )
    );
};

export const createCSV = (
    data: Record<string, any>[],
    columns: TableHeading[],
    translate: (string) => string
): string =>
    unparse(
        {
            data: convertToCSV(data, columns, translate),
            fields: columns.map(({ label }) => translate(label)),
        },
        {
            delimiter: DELIMITER,
        }
    );

const createBlob: (csv: string) => Blob = (csv) => new Blob([`${BOM}${csv}`], { type: 'text/csv' });

const createAnchor: (Blob) => HTMLAnchorElement = (blob) => {
    const link = document.createElement('a');
    link.style.display = 'none';
    link.setAttribute('href', String(URL.createObjectURL(blob)));
    link.setAttribute('download', `report.csv`);
    return link;
};

export const downloadCSV = (csv: string) => {
    const link = createAnchor(createBlob(csv));
    document.body.appendChild(link);
    link.click();
};

export const exportReport = (data: Record<string, any>[], columns: TableHeading[], translate: (string) => string) =>
    downloadCSV(createCSV(data, columns, translate));

export default exportReport;
