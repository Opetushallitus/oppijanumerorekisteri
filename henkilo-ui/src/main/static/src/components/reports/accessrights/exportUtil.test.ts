import { convertToCSV, createCSV, exportReport } from './exportUtil';

const translate = (x: string) => x;

describe('convertToCSV', () => {
    test.each([
        ['Handles empty input', [], [], []],
        ['Handles empty data', [], [{ key: 'label' }], []],
        ['Handles empty columns', [{ key: 'label' }], [], [{}]],
        ['Maps data to correct columns', [{ key: 'value' }], [{ key: 'key', label: 'label' }], [{ label: 'value' }]],
        [
            'Ignored unwanted values',
            [{ key: 'value', foo: 'bar' }],
            [{ key: 'key', label: 'label' }],
            [{ label: 'value' }],
        ],
        ['Non-existing values are undefined', [{}], [{ key: 'key', label: 'label' }], [{ label: undefined }]],
    ])('%s', (_, data, columns, expected) => expect(convertToCSV(data, columns, translate)).toEqual(expected));
});

describe('createCSV', () => {
    test.each([
        ['Handles empty input', [], []],
        ['Handles empty data', [], [{ key: 'key', label: 'label' }]],
        ['Handles data row', [{ key: 'datarow' }], [{ key: 'key', label: 'label' }]],
        ['Handles multiple data rows', [{ key: 'datarow1' }, { key: 'datarow2' }], [{ key: 'key', label: 'label' }]],
        [
            'Handles multiple columns',
            [{ key1: 'item1', key2: 'item1' }],
            [
                { key: 'key1', label: 'label1' },
                { key: 'key2', label: 'label2' },
            ],
        ],
    ])('%s', (_, data, columns) => expect(createCSV(data, columns, translate)).toMatchSnapshot());
});

describe('exportReport', () => {
    window.URL.createObjectURL = jest.fn();

    afterAll(() => {
        jest.resetAllMocks();
    });

    test('Creates a link and appends it to the document', () => {
        exportReport([], [], translate);
        expect(document.body.lastChild).toMatchSnapshot();
    });
});
