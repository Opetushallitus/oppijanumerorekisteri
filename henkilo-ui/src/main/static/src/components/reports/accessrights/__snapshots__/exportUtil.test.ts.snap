// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCSV Handles data row 1`] = `
"label
datarow"
`;

exports[`createCSV Handles empty data 1`] = `
"label
"
`;

exports[`createCSV Handles empty input 1`] = `""`;

exports[`createCSV Handles multiple columns 1`] = `
"label1;label2
item1;item1"
`;

exports[`createCSV Handles multiple data rows 1`] = `
"label
datarow1
datarow2"
`;

exports[`exportReport Creates a link and appends it to the document 1`] = `
<a
  download="report.csv"
  href="undefined"
  style="display: none;"
/>
`;
